### PR TITLE
Added endpoint to list agents

### LINKED ISSUE #3380: Create v2 endpoint to list agents
Create a typed v2 endpoint to list all agents on the server for the provided environment at `/v2/agents` This endpoint is part of the agentmanager.

For each agent the following information should be returned (Agent data class):

- name: The name of the agent
- process: an object that list the name and the uuid of the process. Should be available through id_primary
- last_failover date
- paused
- unpause_on_resume
- status using get_status (not in the database)

the current code sets "" when a value is none, this should be a null in the v2 api so that this is cleaned up a bit.

This endpoint need paging and sorting. It should also support filters on:

- name
- process name
- status

Status might be a bit difficult because the field is not present in the db. TBD

### PR DESCRIPTION
**Pull request by andraskvr**
# Description

closes #3380 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: tests/server/test_agents_api.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        ("process_name", "DESC"),
+        ("process_name", "ASC"),
+        ("last_failover", "DESC"),
+        ("last_failover", "ASC"),
+        ("paused", "DESC"),
+        ("paused", "ASC"),
+        ("unpause_on_resume", "DESC"),
+        ("unpause_on_resume", "ASC"),
+    ],
+)
```
Comment (arnaudsjs) [inline tests/server/test_agents_api.py:132]: You can stack the `pytest.mark.parametrize` annotations to obtain all the combination:
```suggestion
@pytest.mark.parametrize("order_by_column", ["name", "status", "process_name", "last_failover", "paused", "unpause_on_resume"])
@pytest.mark.parametrize("order", ["DESC", "ASC"])
```

File: tests/server/test_agents_api.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+import pytest
+from tornado.httpclient import AsyncHTTPClient, HTTPRequest
+
+from inmanta import data
+from inmanta.server.config import get_bind_port
+
+
+@pytest.fixture
+async def env_with_agents(client, environment):
```
Comment (arnaudsjs) [inline tests/server/test_agents_api.py:32]:
```suggestion
async def env_with_agents(client, environment: str) -> None:
```

File: tests/server/test_agents_api.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    await create_agent(name="paused_agent_with_instance", paused=True, with_process=True)  # paused
+    await create_agent(name="unpause_on_resume", unpause_on_resume=True)  # down
+    await create_agent(
+        name="failover1", with_process=True, last_failover=(datetime.datetime.now() - datetime.timedelta(minutes=1))
+    )  # up
+    await create_agent(name="failover2", with_process=True, last_failover=datetime.datetime.now())  # up
+
+
+@pytest.mark.asyncio
+async def test_agent_list_filters(client, environment, env_with_agents):
```
Comment (arnaudsjs) [inline tests/server/test_agents_api.py:71]:
```suggestion
async def test_agent_list_filters(client, environment: str, env_with_agents: None) -> None:
```

File: tests/server/test_agents_api.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    )
+    assert result.code == 200
+    assert len(result.result["data"]) == 7
+    assert agent_names(result.result["data"]) == all_agent_names_in_expected_order
+
+    assert result.result["metadata"] == {"total": 7, "before": 0, "after": 0, "page_size": 100}
+
+
+@pytest.mark.asyncio
+async def test_sorting_validation(server, client, environment, env_with_agents):
```
Comment (arnaudsjs) [inline tests/server/test_agents_api.py:235]:
```suggestion
async def test_sorting_validation(server, client, environment: str, env_with_agents: None) -> None:
```

File: tests/server/test_agents_api.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        ("last_failover", "DESC"),
+        ("last_failover", "ASC"),
+        ("paused", "DESC"),
+        ("paused", "ASC"),
+        ("unpause_on_resume", "DESC"),
+        ("unpause_on_resume", "ASC"),
+    ],
+)
+@pytest.mark.asyncio
+async def test_agents_paging(server, client, env_with_agents, environment, order_by_column, order):
```
Comment (arnaudsjs) [inline tests/server/test_agents_api.py:134]:
```suggestion
async def test_agents_paging(server, client, env_with_agents: None, environment: str, order_by_column: str, order: str) -> None:
```

File: tests/server/test_agents_api.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    assert result.code == 200
+    assert len(result.result["data"]) == 2
+    assert all(["failover" in agent["name"] for agent in result.result["data"]])
+
+    result = await client.get_agents(environment, filter={"name": "agent", "status": "down"})
+    assert result.code == 200
+    assert len(result.result["data"]) == 1
+
+
+def agent_names(agents):
```
Comment (arnaudsjs) [inline tests/server/test_agents_api.py:112]:
```suggestion
def agent_names(agents: Dict[str, object]) -> List[str]:
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: tests/server/test_agents_api.py (hunk 1 above)
Comment (andraskvr) [reply tests/server/test_agents_api.py]: That's very nice, I didn't know that. Thank you :+1:

File: src/inmanta/data/model.py [hunk 7]
```diff
@@ -453,3 +453,25 @@ class Parameter(BaseModel):
...
     metadata: Optional[JsonType]
+
+
+class Agent(BaseModel):
+    """
+    :param environment: Id of the agent's environment
+    :param name: The name of the agent
+    :param last_failover: The time of the last failover
+    :param paused: Whether the agent is paused or not
+    :param unpause_on_resume: Whether the agent should be unpaused when the environment is resumed
```
[the remark this answers is not in the document]
Comment (andraskvr) [reply src/inmanta/data/model.py:464]: I'm not sure either. @bartv is it useful to return the `unpause_on_resume` field?

[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/data/model.py:464]: @wouterdb wanted it in there iirc.

File: src/inmanta/data/__init__.py [hunk 8]
```diff
@@ -198,6 +205,23 @@ def get_order(self) -> PagingOrder:
         """ The order string representing the direction the results should be sorted by"""
         return self.order
 
+    def is_nullable_column(self) -> bool:
+        """Is the current order by column type nullable (optional) or not"""
+        column_type = self.get_order_by_column_type()
+        return typing_inspect.is_optional_type(column_type)
+
+    def coalesce_to_min(self, value_reference: str) -> ColumnNameStr:
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:213]: Why no use `ORDER BY ... NULLS  FIRST` ?

File: src/inmanta/data/model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/data/model.py:464]: it is usefull to know what will happen when you disengage the emergency stop, so I do like it in here.

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py (hunk 8 above)
Comment (andraskvr) [reply src/inmanta/data/__init__.py:213]: I use that for ordering, but for comparisons, that are needed for paging, the value has to be an actual value. Comparison with `NULL` always yields `NULL`.

File: src/inmanta/data/model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (andraskvr) [reply src/inmanta/data/model.py:464]: Ok, then we will keep it

**Review by arnaudsjs — APPROVED** [review]

File: src/inmanta/data/__init__.py (hunk 8 above)
Comment (wouterdb) [reply src/inmanta/data/__init__.py:213]: I would document this.

File: src/inmanta/protocol/methods_v2.py [hunk 9]
```diff
@@ -348,6 +348,40 @@ def all_agents_action(tid: uuid.UUID, action: AgentAction) -> None:
...
+    :param first_id: The name to use as a continuation token for paging, in combination with the 'start' value,
+            because the order by column might contain non-unique values
+    :param last_id: The name to use as a continuation token for paging, in combination with the 'end' value,
+            because the order by column might contain non-unique values
+                Only one of 'start' and 'end' should be specified at the same time.
+    :param end: The upper limit for the order by column (exclusive).
+                Only one of 'start' and 'end' should be specified at the same time.
+    :param filter: Filter the list of returned agents.
+                Filtering by 'name', 'process_name' and 'status' is supported.
+    :param sort: Return the results sorted according to the parameter value.
```
Comment (wouterdb) [inline src/inmanta/protocol/methods_v2.py:376]: I would not allow sorting on `unpause_on_resume` to much a corner case

**Review by wouterdb — CHANGES_REQUESTED** [review]

**Review by wouterdb — APPROVED** [review]

**Pull request closed without merging**